contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1057 @@
+            "BucketLender#withdraw: The lengths of the input arrays must match"
+        );
+
+        rebalanceBuckets();
+
+        uint256 totalOwedToken = 0;
+        uint256 totalHeldToken = 0;
+
+        uint256 maxHeldToken =
+            Margin(DYDX_MARGIN).isPositionClosed(POSITION_ID) ?
+            TokenInteract.balanceOf(HELD_TOKEN, address(this)) :
+            0;
+
+        for (uint256 i = 0; i < buckets.length; i++) {
+            (uint256 owedTokenForBucket, uint256 heldTokenForBucket) = withdrawInternal(
+                buckets[i],
+                maxWeights[i],
+                maxHeldToken
+            );
+
+            totalOwedToken = totalOwedToken.add(owedTokenForBucket);
+            totalHeldToken = totalHeldToken.add(heldTokenForBucket);
+        }
+
+        // Transfer share of owedToken
+        TokenInteract.transfer(OWED_TOKEN, beneficiary, totalOwedToken);
+        TokenInteract.transfer(HELD_TOKEN, beneficiary, totalHeldToken);
+
+        return (totalOwedToken, totalHeldToken);
+    }
+
+    // ============ Public State-Changing Functions ============
+
+    /**
+     * Allow anyone to refresh the bucket amounts if part of the position was closed since the last
+     * position increase.
+     */
+    function rebalanceBuckets()
+        public
+    {
+        // if force-closed, don't update the outstanding principal values; they are needed to repay
+        // lenders with heldToken
+        if (wasForceClosed) {
+            return;
+        }
+
+        uint256 marginPrincipal = getCurrentPrincipalFromMargin();
+
+        accountForClose(principalTotal.sub(marginPrincipal));
+
+        assert(principalTotal == marginPrincipal);
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * Updates the state variables at any time. Only does anything after the position has been
+     * closed or partially-closed since the last time this function was called.
+     *
+     * - Increases the available amount in the highest buckets with outstanding principal
+     * - Decreases the principal amount in those buckets
+     *
+     * @param  principalRemoved  Amount of principal closed since the last update
+     */
+    function accountForClose(
+        uint256 principalRemoved
+    )
+        internal
+    {
+        if (principalRemoved == 0) {
+            return;
+        }
+
+        uint256 newRepaidAmount = Margin(DYDX_MARGIN).getTotalOwedTokenRepaidToLender(POSITION_ID);
+        assert(newRepaidAmount.sub(cachedRepaidAmount) >= principalRemoved);
+
+        uint256 principalToSub = principalRemoved;
+        uint256 availableToAdd = newRepaidAmount.sub(cachedRepaidAmount);
+        uint256 criticalBucketTemp = criticalBucket;
+
+        // loop over buckets in reverse order starting with the critical bucket
+        for (
+            uint256 bucket = criticalBucketTemp;
+            principalToSub > 0 && bucket <= criticalBucketTemp; // prevent underflow on bucket
```

Comment:
Oh this `&& bucket <= criticalBucketTemp` was for underflow? that's pretty confusing

Why would this ever underflow? Shouldn't we never go past the first bucket in terms of the amount of principal we're paying back?

Reply:
it shouldn't underflow but assert doesn't have a return value so I can't do 

`principalToSub > 0 && assert(bucket <= criticalBucketTemp)`

Reply:
but I can put the assert as the first statement